components/omega/src/ocn/Tracers.cpp

```diff
@@ -0,0 +1,527 @@
+// Initialization
+//---------------------------------------------------------------------------
+I4 Tracers::init() {
+
+   I4 Err = 0;
+
+   LOG_INFO("Tracers::init() called");
+
+   // Retrieve mesh cell/edge/vertex totals from Decomp
+   HorzMesh *DefHorzMesh = HorzMesh::getDefault();
+
+   NCellsOwned = DefHorzMesh->NCellsOwned;
+   NCellsAll   = DefHorzMesh->NCellsAll;
+   NCellsSize  = DefHorzMesh->NCellsSize;
+   NVertLevels = DefHorzMesh->NVertLevels;
+
+   MeshHalo = Halo::getDefault();
+
+   auto *DefTimeStepper = TimeStepper::getDefault();
+   if (!DefTimeStepper) {
+      LOG_ERROR("TimeStepper needs to be initialized before Tracers");
+      return -1;
+   }
+   NTimeLevels = DefTimeStepper->getNTimeLevels();
+
+   if (NTimeLevels < 2) {
+      LOG_ERROR("Tracers: the number of time level is lower than 2");
+      return -2;
+   }
+
+   CurTimeIndex = 0;
+
+   // load Tracers configs
+   Config *OmegaConfig = Config::getOmegaConfig();
+   Config TracersConfig("Tracers");
+   Err = OmegaConfig->get(TracersConfig);
+   if (Err != 0) {
+      LOG_ERROR("Tracers: Tracers group not found in Config");
+      return -3;
+   }
+
+   NumTracers     = 0;
+   I4 TracerIndex = 0;
+
+   // get tracers group and tracer names
+   for (auto It = TracersConfig.begin(); It != TracersConfig.end(); ++It) {
+
+      I4 GroupStartIndex = TracerIndex;
+
+      std::string GroupName;
+      I4 GroupNameErr = OMEGA::Config::getName(It, GroupName);
+      if (GroupNameErr != 0) {
+         LOG_ERROR("Tracers: {} tracer group name not found in TracersConfig",
+                   GroupName);
+         return -4;
+      }
+
+      std::vector<std::string> _TracerNames;
+      I4 TracerNamesErr = TracersConfig.get(GroupName, _TracerNames);
+      if (TracerNamesErr != 0) {
+         LOG_ERROR("Tracers: {} group tracers not found in TracersConfig",
+                   GroupName);
+         return -5;
+      }
+
+      for (auto _TracerName : _TracerNames) {
+         TracerIndexes[_TracerName] = TracerIndex;
+         TracerIndex++;
+      }
+
+      TracerGroups[GroupName] =
+          std::pair<I4, I4>(GroupStartIndex, TracerIndex - GroupStartIndex);
+      FieldGroup::create("TracerGroup" + GroupName);
+   }
+
+   // total number of tracers
+   NumTracers = TracerIndex;
+
+   // Initialize tracers arrays for device and host
+   TracerArrays.resize(NTimeLevels);
+   TracerArraysH.resize(NTimeLevels);
+
+   // Allocate tracers data array and assign to tracers arrays
+   for (I4 TimeIndex = 0; TimeIndex < NTimeLevels; ++TimeIndex) {
+      TracerArrays[TimeIndex] =
+          Array3DReal("TracerTime" + std::to_string(TimeIndex), NumTracers,
+                      NCellsSize, NVertLevels);
+      TracerArraysH[TimeIndex] =
+          HostArray3DReal("TracerHTime" + std::to_string(TimeIndex), NumTracers,
+                          NCellsSize, NVertLevels);
+   }
+
+// Read tracer definitions from file
+#include "TracerDefs.inc"
```

Comment:
@philipwjones is your suggestion somewhat equivalent to taking this part of the init routine and adding it to the include file?  Or are you proposing to not iterate over the groups in the tracer config and just having hard coded indices for each tracer no matter if they are active or not?

Reply:
@vanroekel Not hard-coded indices. We would still initialize at run time, just provide a shared named index that we could refer to rather than retrieving by name. But it looks like that's going to be more complicated than I thought so I think we'll stick with the current implementation for now.

Reply:
Got it, so mostly moving the init to the .inc file, but I did see that Youngsung found it doesn't work as easily as possible.  Thanks.